System/Random.hs

```diff
@@ -68,61 +215,91 @@
+-- with 'RandomGen' instances based on 'next' and 'genRange'. This means that
+-- they are not optimal for pseudo-random number generators with a power-of-2
+-- modulo.
 --
--- Instead we should define (where e.g. @unBuildWord32 :: Word32 ->
--- (Word16, Word16)@ is a function to pull apart a 'Word32' into a
--- pair of 'Word16'):
+-- So let's implement a faster 'RandomGen' instance for our pseudo-random
+-- number generator as follows:
 --
 -- >>> newtype PCGen' = PCGen' { unPCGen :: PCGen }
---
 -- >>> let stepGen' = second PCGen' . stepGen . unPCGen
---
 -- >>> :{
 -- instance RandomGen PCGen' where
 --   genWord8 = first fromIntegral . stepGen'
 --   genWord16 = first fromIntegral . stepGen'
 --   genWord32 = stepGen'
 --   genWord64 g = (buildWord64 x y, g'')
---       where
+--     where
 --       (x, g') = stepGen' g
 --       (y, g'') = stepGen' g'
---       buildWord64 w0 w1 = ((fromIntegral w1) `shiftL` 32) .|. (fromIntegral w0)
 -- :}
 --
--- [/Example for RNG Users:/]
+-- == How to implement 'RandomGen' for a pseudo-random number generator without a power-of-2 modulus
 --
--- Suppose you want to simulate rolls from a dice (yes I know it's a
--- plural form but it's now common to use it as a singular form):
+-- Suppose you want to implement the pseudo-random number generator from
+-- <https://dl.acm.org/doi/abs/10.1145/62959.62969 L’Ecuyer (1988)>. It
+-- natively generates an integer value in the range @[1, 2147483562]@. This is
+-- the generator used by this library before it was replaced by SplitMix in
+-- version 1.2.
 --
+-- >>> data LegacyGen = LegacyGen !Int32 !Int32
 -- >>> :{
--- let randomListM :: (MonadRandom g m, Num a, Uniform a) => g -> Int -> m [a]
---     randomListM gen n = replicateM n (uniform gen)
+-- let legacyNext :: LegacyGen -> (Int, LegacyGen)
+--     legacyNext (LegacyGen s1 s2) = (fromIntegral z', LegacyGen s1'' s2'') where
+--       z' = if z < 1 then z + 2147483562 else z
+--       z = s1'' - s2''
+--       k = s1 `quot` 53668
+--       s1'  = 40014 * (s1 - k * 53668) - k * 12211
+--       s1'' = if s1' < 0 then s1' + 2147483563 else s1'
+--       k' = s2 `quot` 52774
+--       s2' = 40692 * (s2 - k' * 52774) - k' * 3791
+--       s2'' = if s2' < 0 then s2' + 2147483399 else s2'
 -- :}
 --
+-- You can make it an instance of 'RandomGen' as follows:
+--
 -- >>> :{
--- let rolls :: [Word32]
---     rolls = runGenState_
---               (PCGen 17 29)
---               (\g -> randomListM g 10 >>= \xs -> return $ map ((+1) . (`mod` 6)) xs)
+-- instance RandomGen LegacyGen where
+--   next = legacyNext
+--   genRange _ = (1, 2147483562)
 -- :}
 --
--- >>> rolls
--- [1,4,2,4,2,2,3,1,5,1]
+-- Note that since the default implementations of all other 'RandomGen' methods
+-- are geared towards pseudo-random number generators with 'next' as the source
+-- of randomness for backwards compatibility, there is no need to implement
+-- additional methods. You should of course verify this with a benchmark.
```

Comment:
I think we need to emphasise that if a user implements only `next` and `genRange` the whole thing will be stupid slow.

> all other 'RandomGen' methods are geared towards pseudo-random number generators with 'next' as the source of randomness for backwards compatibility

this can be read as "`next` is the only thing you need to do and all other methods will fall in place", which isn't accurate. `next` and `genRange` are provided for backwards compatibility only. The best thing would be is to say something along the lines: this was the "old way" and the "new way" is you need to implement either `genWord32` or `genWord64` and only then you RNG will perform fast.

Reply:
I've attempted to address this here: https://github.com/idontgetoutmuch/random/pull/61/commits/99f47a2083c97e8dea2229c82027f97ff6a22bce